In [1]:
import pandas as pd
import re
import numpy as np
import string
import json 
from pprint import pprint
import progressbar
import openpyxl as pyxl

# JSON values:

In [2]:
json_df = pd.read_csv(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\Sample_json.csv")
json_df= json_df["_raw"]
json_df = json_df.dropna()

C:\Users\kmoudgalya\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2728: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,24

In [3]:
# for key, value in json_df.iteritems():
#     print(value)

# Finding keys in the original JSON log file

In [4]:
found_keys = []

regex_keys = r"\"[a-zA-Z0-9._\\]+\":"

for key, value in json_df.iteritems():
    matches = re.findall(regex_keys, str(value))
    for j in matches:
        if j not in found_keys:
            found_keys.append(str(j)[1:])

#     if matches not in found_keys:
#         #found_keys.append(str(matches))
#         found_keys.append(matches)

In [5]:
len(found_keys)

97825

### Set() eliminates duplicates, then we check the count.

In [6]:
len(set(found_keys))

396

In [7]:
from collections import Counter, OrderedDict
print(Counter(found_keys).keys())
print(Counter(found_keys).values())
print(len(Counter(found_keys).keys()))
print(len(Counter(found_keys).values()))

dict_keys(['timestamp":', 'bytes_in":', 'bytes_out":', 'bytes":', 'dest_ip":', 'dest_port":', 'time_taken":', 'transport":', 'src_ip":', 'site":', 'http_method":', 'status":', 'url":', 'http_user_agent":', 'src_content":', 'orderid\\":', 'dest_content":', 'orderstatus\\":', 'userid\\":', 'cartid\\":', 'item\\":', 'items\\":', 'user":', 'query":', 'get_all_statistics":', 'pool_name":', 'statistics":', 'time_stamp":', 'type":', 'value":', 'high":', 'low":', 'f5_bigip_partition_name":', 'mlt_type":', 'with_field":', 'break_field":', 'module_interface":', 'get_http_statistics":', 'account_id":', 'application_summary":', 'response_time":', 'throughput":', 'error_rate":', 'apdex_target":', 'apdex_score":', 'host_count":', 'instance_count":', 'end_user_summary":', 'health_status":', 'id":', 'input_name":', 'name":', 'transaction_name":', 'reporting":', 'modelerType":', 'sessionIdLength":', 'className":', 'secureRandomAlgorithm":', 'maxInactiveInterval":', 'secureRandomClass":', 'sessionAverag

In [8]:
for i in Counter(found_keys).keys():
    print(i)

timestamp":
bytes_in":
bytes_out":
bytes":
dest_ip":
dest_port":
time_taken":
transport":
src_ip":
site":
http_method":
status":
url":
http_user_agent":
src_content":
orderid\":
dest_content":
orderstatus\":
userid\":
cartid\":
item\":
items\":
user":
query":
get_all_statistics":
pool_name":
statistics":
time_stamp":
type":
value":
high":
low":
f5_bigip_partition_name":
mlt_type":
with_field":
break_field":
module_interface":
get_http_statistics":
account_id":
application_summary":
response_time":
throughput":
error_rate":
apdex_target":
apdex_score":
host_count":
instance_count":
end_user_summary":
health_status":
id":
input_name":
name":
transaction_name":
reporting":
modelerType":
sessionIdLength":
className":
secureRandomAlgorithm":
maxInactiveInterval":
secureRandomClass":
sessionAverageAliveTime":
rejectedSessions":
processExpiresFrequency":
stateName":
duplicates":
distributable":
maxActiveSessions":
sessionMaxAliveTime":
processingTime":
pathname":
sessionExpireRate":
activeSes

In [9]:
Counter(found_keys)

Counter({'Arch":': 45,
         'Assigned_To":': 2,
         'Assigned_To\\":': 1,
         'AvailableProcessors":': 45,
         'Category":': 2,
         'Category\\":': 1,
         'CollectionUsage":': 90,
         'CollectionUsageThresholdSupported":': 90,
         'CommittedVirtualMemorySize":': 45,
         'Company":': 2,
         'Company\\":': 1,
         'Department":': 2,
         'Department\\":': 1,
         'Description":': 2,
         'Description\\":': 1,
         'FreePhysicalMemorySize":': 45,
         'FreeSwapSpaceSize":': 45,
         'HeapMemoryUsage":': 45,
         'Impact":': 2,
         'Impact\\":': 1,
         'MaxFileDescriptorCount":': 45,
         'MemoryManagerNames":': 90,
         'Name":': 135,
         'NetObject":': 6,
         'NonHeapMemoryUsage":': 45,
         'ObjectName":': 180,
         'ObjectPendingFinalizationCount":': 45,
         'OpenFileDescriptorCount":': 45,
         'Organization":': 2,
         'Organization\\":': 1,
         'Peak

In [10]:
OrderedDict(sorted(Counter(found_keys).items(), key=lambda t: t[1], reverse = True))

OrderedDict([('value":', 3244),
             ('max":', 2911),
             ('startTimeInMillis":', 2651),
             ('frequency":', 2642),
             ('metricId":', 2642),
             ('metricName":', 2642),
             ('metricPath":', 2642),
             ('metricValues":', 2641),
             ('count":', 2641),
             ('current":', 2641),
             ('min":', 2636),
             ('occurrences":', 2636),
             ('standardDeviation":', 2636),
             ('sum":', 2635),
             ('useRange":', 2635),
             ('timestamp":', 1119),
             ('name":', 804),
             ('bytes":', 776),
             ('dest_ip":', 776),
             ('dest_port":', 776),
             ('time_taken":', 776),
             ('transport":', 776),
             ('src_ip":', 776),
             ('response_time":', 762),
             ('throughput":', 762),
             ('apdex_target":', 762),
             ('apdex_score":', 762),
             ('status":', 703),
             ('id

# Find out number of unique elements. If there are more than 50/100 elements, take the top 50/100 and
# use those as columns for these kinds of logs

# We have 396 unique columns, should probably be fine, expect long execution times, remember to use progressbar2

In [11]:
col_names = []
for j in Counter(found_keys).keys():
    if j[-3] != "\\":
        col_names.append(j)
    else:
        col_names.append(j+"##")

print(col_names)
#         if j.endswith("\":"):
#             col_names.append(j.replace("\":","##"))
#         print(j.replace("\":","##"))
# for j in Counter(found_keys).keys():
#     print(j)

['timestamp":', 'bytes_in":', 'bytes_out":', 'bytes":', 'dest_ip":', 'dest_port":', 'time_taken":', 'transport":', 'src_ip":', 'site":', 'http_method":', 'status":', 'url":', 'http_user_agent":', 'src_content":', 'orderid\\":##', 'dest_content":', 'orderstatus\\":##', 'userid\\":##', 'cartid\\":##', 'item\\":##', 'items\\":##', 'user":', 'query":', 'get_all_statistics":', 'pool_name":', 'statistics":', 'time_stamp":', 'type":', 'value":', 'high":', 'low":', 'f5_bigip_partition_name":', 'mlt_type":', 'with_field":', 'break_field":', 'module_interface":', 'get_http_statistics":', 'account_id":', 'application_summary":', 'response_time":', 'throughput":', 'error_rate":', 'apdex_target":', 'apdex_score":', 'host_count":', 'instance_count":', 'end_user_summary":', 'health_status":', 'id":', 'input_name":', 'name":', 'transaction_name":', 'reporting":', 'modelerType":', 'sessionIdLength":', 'className":', 'secureRandomAlgorithm":', 'maxInactiveInterval":', 'secureRandomClass":', 'sessionAver

In [12]:
str('Company\\":##')

'Company\\":##'

## timestamp": and timestamp\\": are counted as different entries in the Collections dict, we need to fix that here

## ACTUALLY NO.

## A data cell may have both timestamp: and timestamp\: both in the same cell, check number 3244 for proof

In [13]:
#col_names

In [14]:
# output_v1 = json_df
# output_v1.columns = col_names
# output_v1.to_excel(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\trial_1.xlsx")

In [15]:
pbar = progressbar.ProgressBar()

In [16]:
a = 'Company\\":##'
a[:-5]

'Company'

## experimenting to get / ones

In [17]:
#pbar = progressbar.ProgressBar()
output_v1 = json_df

for i in col_names:
    if i.endswith("##") and i.startswith("orderid"):
        #print("inside this")
        #regex = (i[:-5]) + "\\\": \\\"[0-9a-zA-Z- :/?=!\.]+\\\""
        regex = i[:-5]  +   "..: ..[0-9a-zA-Z- :/?=!\.]+.."
        #print("Regex is: {}".format(regex))
        temp_list = []
        for key,value in json_df.iteritems():
            matches = re.findall(regex, str(value))
            temp_list.append(str(matches)[len(i)+3:-2])
        data = np.array((temp_list))
        temp_df = pd.DataFrame(data)
        output_v1 = pd.concat([output_v1,temp_df],axis = 1)     

In [18]:
str(regex)

'orderid..: ..[0-9a-zA-Z- :/?=!\\.]+..'

# Final JSON data extractor

## This one works, dont use, use the one below this

In [19]:
# output_v1 = json_df
# pbar = progressbar.ProgressBar()

# for i in pbar(col_names):
#     if i.endswith("##"):
#         regex = "\"" + i[:-5]  +   "..: ..[0-9a-zA-Z- :/?=!\.]+.."
#         temp_list = []
#         for key,value in json_df.iteritems():
#             matches = re.findall(regex, str(value))
#             temp_list.append(str(matches)[len(i)+3:-2])
#         data = np.array((temp_list))
#         temp_df = pd.DataFrame(data)
#         output_v1 = pd.concat([output_v1,temp_df],axis = 1) 
#     else:
#         regex = "\"" + i + " \"[0-9a-zA-Z- :/?=!\.]+\"|" + "\"" + i + " [0-9a-zA-Z- :/?=!\.]+,|" + "\"" + i[:-2] + "=\\\"[0-9a-zA-Z- :/?=!\.]+\\\""
#     #     else:
#     #         regex = i + " \\\"[0-9a-zA-Z- :/?=!\.]+\\\""
#     #     print("Regex is: {}".format(regex))
#     #     if i == "bytes_in":
#     #         print("Breaking on i = 2. Check output now.")
#     #         break
#     #     else:
#         temp_list = []
#         for key,value in json_df.iteritems():
#             matches = re.findall(regex, str(value))
#             temp_list.append(str(matches)[len(i)+3:-2])
#         data = np.array((temp_list))
#         temp_df = pd.DataFrame(data)
#         output_v1 = pd.concat([output_v1,temp_df],axis = 1)     

# modified above, to sovle the 3244 value": mismatches

In [50]:
output_v1 = json_df
pbar = progressbar.ProgressBar()

for i in pbar(col_names):
    if i.endswith("##"):
        regex = "\"" + i[:-5]  +   "..: ..[0-9a-zA-Z- :/?=!\.]+.."
        temp_list = []
        for key,value in json_df.iteritems():
            matches = re.findall(regex, str(value))
            #temp_list.append(str(matches)[len(i)+3:-2])
            temp_list.append(str(matches)[2:-1])
        data = np.array((temp_list))
        temp_df = pd.DataFrame(data)
        output_v1 = pd.concat([output_v1,temp_df],axis = 1) 
    elif i == 'value":':
        regex = "\"" + i  +   " {\"high\": [0-9]+, \"low\": [0-9]+}}|" + "\"" + i[-2] + "\"\": [0-9]+"
        temp_list = []
        for key,value in json_df.iteritems():
            matches = re.findall(regex, str(value))
            #temp_list.append(str(matches)[len(i)+3:-2])
            temp_list.append(str(matches)[2:-1])
        data = np.array((temp_list))
        temp_df = pd.DataFrame(data)
        output_v1 = pd.concat([output_v1,temp_df],axis = 1) 
    else:
        regex = "\"" + i + " \"[0-9a-zA-Z- :/?=!\.]+\"|" + "\"" + i + " [0-9a-zA-Z- :/?=!\.]+,|" + "\"" + i[:-2] + "=\\\"[0-9a-zA-Z- :/?=!\.]+\\\""
    #     else:
    #         regex = i + " \\\"[0-9a-zA-Z- :/?=!\.]+\\\""
    #     print("Regex is: {}".format(regex))
    #     if i == "bytes_in":
    #         print("Breaking on i = 2. Check output now.")
    #         break
    #     else:
        temp_list = []
        for key,value in json_df.iteritems():
            matches = re.findall(regex, str(value))
            #temp_list.append(str(matches)[len(i)+3:-2])
            temp_list.append(str(matches)[2:-1])
        data = np.array((temp_list))
        temp_df = pd.DataFrame(data)
        output_v1 = pd.concat([output_v1,temp_df],axis = 1)     

100% (396 of 396) |######################| Elapsed Time: 0:00:09 Time:  0:00:09


In [51]:
#=\\\"[0-9a-zA-Z- :/?=!\.]+\\\"
a = "=\\\"[0-9a-zA-Z- :/?=!\.]+\\\""
print(a)

=\"[0-9a-zA-Z- :/?=!\.]+\"


In [52]:
output_v1.columns = ["raw"] + col_names

In [53]:
output_v1

raw  \
0     {"timestamp": "2018-10-23 15:02:19.115992", "b...   
1     {"timestamp": "2018-10-23 15:02:19.115981", "b...   
2     {"timestamp": "2018-10-23 15:02:19.115969", "b...   
3     {"timestamp": "2018-10-23 15:02:19.115963", "b...   
4     {"timestamp": "2018-10-23 15:02:19.115961", "b...   
5     {"timestamp": "2018-10-23 15:02:19.115950", "b...   
6     {"timestamp": "2018-10-23 15:02:19.115941", "b...   
7     {"timestamp": "2018-10-23 15:02:19.115932", "b...   
8     {"timestamp": "2018-10-23 15:02:19.115930", "b...   
9     {"timestamp": "2018-10-23 15:02:19.115914", "b...   
10    {"timestamp": "2018-10-23 15:02:19.115914", "b...   
11    {"timestamp": "2018-10-23 15:02:19.115913", "b...   
12    {"timestamp": "2018-10-23 15:02:19.115910", "b...   
13    {"timestamp": "2018-10-23 15:02:19.115907", "b...   
14    {"timestamp": "2018-10-23 15:02:19.115906", "b...   
15    {"timestamp": "2018-10-23 15:02:19.115891", "b...   
16    {"timestamp": "2018-10-23 15:02:19.115890", "b...   
17    {"timestamp": "2018-10-23 15:02:19.115887", "b...   
18    {"timestamp": "2018-10-23 15:02:19.115878", "b...   
19    {"timestamp": "2018-10-23 15:02:19.115835","us...   
20    {"timestamp": "2018-10-23 15:02:19.115818","us...   
21    {"timestamp": "2018-10-23 15:02:19.115809", "b...   
22    {"timestamp": "2018-10-23 15:02:19.115807","us...   
23    {"timestamp": "2018-10-23 15:02:19.115793","us...   
24    {"timestamp": "2018-10-23 15:02:19.115785","us...   
25    {"timestamp": "2018-10-23 15:02:19.115763","us...   
26    {"timestamp": "2018-10-23 15:02:19.115758","us...   
27    {"timestamp": "2018-10-23 15:02:19.115747","us...   
28    {"timestamp": "2018-10-23 15:02:19.115732","us...   
29    {"timestamp": "2018-10-23 15:02:19.115721","us...   
...                                                 ...   
3237  {"title": "Puppet Event: 3b2f217637430002ad65b...   
3238  {"title": "Puppet Event: 3b2f217637430002ad65b...   
3239  {"title": "Puppet Event: 3b2f217637430002ad65b...   
3240  {"is_service_aggregate": "0", "is_use_event_ti...   
3241  {"is_service_aggregate": "0", "is_use_event_ti...   
3242  {"is_service_aggregate": "0", "is_use_event_ti...   
3243  {"is_service_aggregate": "0", "time": "600", "...   
3244  {"is_service_aggregate": "0", "is_use_event_ti...   
3245  {"is_service_aggregate": "0", "time": "600", "...   
3246  {"is_service_aggregate": "0", "is_use_event_ti...   
3247  {"is_service_aggregate": "0", "time": "600", "...   
3248  {"is_service_aggregate": "0", "is_use_event_ti...   
3249  {"is_service_aggregate": "0", "time": "600", "...   
3250  {"is_service_aggregate": "0", "is_use_event_ti...   
3251  {"is_service_aggregate": "0", "time": "600", "...   
3252  {"is_service_aggregate": "0", "is_use_event_ti...   
3253  {"is_service_aggregate": "0", "time": "600", "...   
3254  {"is_service_aggregate": "0", "is_use_event_ti...   
3255  {"is_service_aggregate": "0", "time": "600", "...   
3256  {"is_service_aggregate": "0", "is_use_event_ti...   
3257  {"is_service_aggregate": "0", "time": "600", "...   
3258  {"is_service_aggregate": "0", "is_use_event_ti...   
3259  {"is_service_aggregate": "0", "time": "600", "...   
3260  {"is_service_aggregate": "0", "is_use_event_ti...   
3261  {"is_service_aggregate": "0", "time": "600", "...   
3262  {"is_service_aggregate": "0", "is_use_event_ti...   
3263  {"is_service_aggregate": "0", "time": "600", "...   
3264  {"is_service_aggregate": "0", "is_use_event_ti...   
3265  {"is_service_aggregate": "0", "time": "600", "...   
3266  {"drilldown_search_earliest_offset": "-1800", ...   

                                     timestamp":          bytes_in":  \
0     "timestamp": "2018-10-23 15:02:19.115992"'   "bytes_in": 735,'   
1     "timestamp": "2018-10-23 15:02:19.115981"'   "bytes_in": 450,'   
2     "timestamp": "2018-10-23 15:02:19.115969"'   "bytes_in": 420,'   
3     "timestamp": "2018-10-23 15:02:19.115963"'   "bytes_in": 561,'   
4     "timestamp": "2018-10-23 15

In [54]:
output_v1.to_csv((r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\trial_1.csv"))

# Checking the output:

In [55]:
check_file = OrderedDict(sorted(Counter(found_keys).items(), key=lambda t: t[1], reverse = True))
check_df = pd.DataFrame(list(check_file.items()))
yes_or_no = np.zeros((396,1))
yes_no_df = pd.DataFrame(yes_or_no)

In [56]:
check_file

OrderedDict([('value":', 3244),
             ('max":', 2911),
             ('startTimeInMillis":', 2651),
             ('frequency":', 2642),
             ('metricId":', 2642),
             ('metricName":', 2642),
             ('metricPath":', 2642),
             ('metricValues":', 2641),
             ('count":', 2641),
             ('current":', 2641),
             ('min":', 2636),
             ('occurrences":', 2636),
             ('standardDeviation":', 2636),
             ('sum":', 2635),
             ('useRange":', 2635),
             ('timestamp":', 1119),
             ('name":', 804),
             ('bytes":', 776),
             ('dest_ip":', 776),
             ('dest_port":', 776),
             ('time_taken":', 776),
             ('transport":', 776),
             ('src_ip":', 776),
             ('response_time":', 762),
             ('throughput":', 762),
             ('apdex_target":', 762),
             ('apdex_score":', 762),
             ('status":', 703),
             ('id

In [58]:
# check_df = pd.concat([check_df, yes_no_df], axis = 1)
# check_df.columns = ["Check Column","Expected Value", "Match? 1 or 0, default 0"]
# check_df.to_csv(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\check_json_output.csv")

In [59]:
from collections import Counter, defaultdict
import csv

# header_counter = defaultdict(Counter)
# count_dict = {}
# with open(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\trial_1.csv") as input_file:
#     r = csv.reader(input_file, delimiter=',')
#     # read headers
#     headers = next(r)
#     headers_1 = r.fieldnames
#     count_list = []
#     for row in r:
#         count = 0
#         if row != "":
#             count = count + 1
def get_number_of_rows(filename, header):
      with open(filename) as f:
         reader = csv.DictReader(f)
         return sum([bool(row[header]) for row in reader])

In [60]:
reader = csv.DictReader(open(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\trial_1.csv"))
reader.fieldnames

['',
 'raw',
 'timestamp":',
 'bytes_in":',
 'bytes_out":',
 'bytes":',
 'dest_ip":',
 'dest_port":',
 'time_taken":',
 'transport":',
 'src_ip":',
 'site":',
 'http_method":',
 'status":',
 'url":',
 'http_user_agent":',
 'src_content":',
 'orderid\\":##',
 'dest_content":',
 'orderstatus\\":##',
 'userid\\":##',
 'cartid\\":##',
 'item\\":##',
 'items\\":##',
 'user":',
 'query":',
 'get_all_statistics":',
 'pool_name":',
 'statistics":',
 'time_stamp":',
 'type":',
 'value":',
 'high":',
 'low":',
 'f5_bigip_partition_name":',
 'mlt_type":',
 'with_field":',
 'break_field":',
 'module_interface":',
 'get_http_statistics":',
 'account_id":',
 'application_summary":',
 'response_time":',
 'throughput":',
 'error_rate":',
 'apdex_target":',
 'apdex_score":',
 'host_count":',
 'instance_count":',
 'end_user_summary":',
 'health_status":',
 'id":',
 'input_name":',
 'name":',
 'transaction_name":',
 'reporting":',
 'modelerType":',
 'sessionIdLength":',
 'className":',
 'secureRandomAlgo

# The cell below will take around 3 minutes to execute

In [61]:
reader = csv.DictReader(open(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\trial_1.csv"))
col_headers = reader.fieldnames
count_dict = {}
pbar = progressbar.ProgressBar()
for i in pbar(range(2,len(col_headers))):
    count = get_number_of_rows(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\trial_1.csv",col_headers[i])
    count_dict[col_headers[i]] = count

100% (396 of 396) |######################| Elapsed Time: 0:02:09 Time:  0:02:09


In [62]:
#count_dict
count_dict2 = OrderedDict(sorted(Counter(count_dict).items(), key=lambda t: t[1], reverse = True))
count_dict2

OrderedDict([('timestamp":', 1119),
             ('bytes":', 776),
             ('dest_ip":', 776),
             ('dest_port":', 776),
             ('time_taken":', 776),
             ('transport":', 776),
             ('src_ip":', 776),
             ('name":', 709),
             ('status":', 701),
             ('id":', 629),
             ('account_id":', 612),
             ('health_status":', 606),
             ('input_name":', 606),
             ('reporting":', 606),
             ('transaction_name":', 518),
             ('bytes_in":', 410),
             ('bytes_out":', 410),
             ('site":', 410),
             ('http_method":', 410),
             ('http_user_agent":', 410),
             ('response_time":', 381),
             ('throughput":', 381),
             ('error_rate":', 381),
             ('apdex_target":', 381),
             ('apdex_score":', 381),
             ('host_count":', 381),
             ('instance_count":', 381),
             ('url":', 380),
             ('u

In [63]:
check_df2 = pd.DataFrame(list(count_dict2.items()))

In [64]:
check_df2

0     1
0                         timestamp":  1119
1                             bytes":   776
2                           dest_ip":   776
3                         dest_port":   776
4                        time_taken":   776
5                         transport":   776
6                            src_ip":   776
7                              name":   709
8                            status":   701
9                                id":   629
10                       account_id":   612
11                    health_status":   606
12                       input_name":   606
13                        reporting":   606
14                 transaction_name":   518
15                         bytes_in":   410
16                        bytes_out":   410
17                             site":   410
18                      http_method":   410
19                  http_user_agent":   410
20                    response_time":   381
21                       throughput":   381
22                       error_rate":   381
23                     apdex_target":   381
24                      apdex_score":   381
25                       host_count":   381
26                   instance_count":   381
27                              url":   380
28                             user":   370
29                   jvmDescription":   352
..                                ...   ...
366          application_summary\":##     0
367             end_user_summary\":##     0
368                           alert":     0
369                network_in_bytes":     0
370                          reason":     0
371                  object_creator":     0
372                   audit_creator":     0
373                             uri":     0
374                         content":     0
375                     severity\":##     0
376                        owner\":##     0
377                     group_id\":##     0
378                  description\":##     0
379                        title\":##     0
380                        activity":     0
381                  transaction_id":     0
382                    exec_message":     0
383         itsi_dummy_closing_flag":     0
384              application_events":     0
385                affectedEntities":     0
386                         subType":     0
387                         summary":     0
388                 triggeredEntity":     0
389      itsi_group_ace_text_values":     0
390  itsi_group_ace_category_values":     0
391                break_group_flag":     0
392                       old_value":     0
393                       new_value":     0
394                    old_value\":##     0
395                    new_value\":##     0

[396 rows x 2 columns]

In [65]:
check_df3 = pd.concat([check_df, yes_no_df, check_df2], axis = 1)
check_df3.columns = ["Check Column","Expected Value", "Match? 1 or 0, default 0", "Check Column", "Actual Value"]
check_df3.to_csv(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\check_json_output5.csv")

# Make sure the cell counter is working properly, use it on a smaller excel file to verify

# If it is working properly, look into regexes

# If the counter is not working properly, figure out how to fix the counter

## Checkign the output v2:

In [66]:
check_file = OrderedDict(sorted(Counter(found_keys).items(), key=lambda t: t[1], reverse = True))
check_df = pd.DataFrame(list(check_file.items()))
yes_or_no = np.zeros((396,1))
yes_no_df = pd.DataFrame(yes_or_no)

In [67]:
found_keys_new = []

regex_keys = r"\"[a-zA-Z0-9._\\]+\":"#|\"[a-zA-Z0-9._\\]+\":##" 
count = 0
for key, value in output_v1.iteritems():
    matches = re.findall(regex_keys, str(value))
    for j in matches:
            found_keys_new.append(str(j)[1:])

In [68]:
print(len(found_keys_new))
print(len(set(found_keys_new)))

1616
119


In [69]:
from collections import Counter, OrderedDict
print(Counter(found_keys_new).keys())
print(Counter(found_keys_new).values())
print(len(Counter(found_keys_new).keys()))
print(len(Counter(found_keys_new).values()))

dict_keys(['timestamp":', 'title":', 'is_service_aggregate":', 'time":', 'drilldown_search_earliest_offset":', 'bytes_in":', 'bytes_out":', 'bytes":', 'dest_ip":', 'dest_port":', 'time_taken":', 'transport":', 'src_ip":', 'site":', 'http_method":', 'status":', 'url":', 'http_user_agent":', 'orderid\\\\":', 'orderstatus\\\\":', 'userid\\\\":', 'cartid\\\\":', 'item\\\\":', 'user":', 'query":', 'description":', 'severity":', 'latency":', 'event_id":', 'search_name":', 'orig_index":', 'service_ids":', 'drilldown_search_search":', 'serviceid":', 'norm_health_status":', 'orig_rid":', 'norm_instance":', 'entity_key":', 'event_identifier_hash":', 'drilldown_uri":', 'is_entity_defined":', 'norm_severity":', 'owner":', 'drilldown_search_latest_offset":', 'is_use_event_time":', 'drilldown_search_title":', 'rid":', 'orig_sourcetype":', 'drilldown_title":', 'mod_time":', 'is_entity_in_maintenance":', 'entity_title":', 'search_type":', 'orig_status":', 'norm_message":', 'executiontime":', 'src_host

In [70]:
OrderedDict(sorted(Counter(found_keys_new).items(), key=lambda t: t[1], reverse = True))

OrderedDict([('timestamp":', 63),
             ('is_service_aggregate":', 55),
             ('status":', 50),
             ('drilldown_search_earliest_offset":', 31),
             ('bytes":', 30),
             ('dest_ip":', 30),
             ('dest_port":', 30),
             ('time_taken":', 30),
             ('transport":', 30),
             ('src_ip":', 30),
             ('severity":', 30),
             ('event_id":', 30),
             ('search_name":', 30),
             ('orig_index":', 30),
             ('orig_rid":', 30),
             ('event_identifier_hash":', 30),
             ('drilldown_uri":', 30),
             ('norm_severity":', 30),
             ('owner":', 30),
             ('drilldown_search_latest_offset":', 30),
             ('is_use_event_time":', 30),
             ('drilldown_search_title":', 30),
             ('rid":', 30),
             ('orig_sourcetype":', 30),
             ('drilldown_title":', 30),
             ('mod_time":', 30),
             ('search_type":',

In [41]:
check_file2 = OrderedDict(sorted(Counter(found_keys_new).items(), key=lambda t: t[1], reverse = True))
check_df2 = pd.DataFrame(list(check_file2.items()))

In [42]:
check_df3 = pd.concat([check_df, yes_no_df, check_df2], axis = 1)
check_df3.columns = ["Check Column","Expected Value", "Match? 1 or 0, default 0", "Check Column", "Actual Value"]
check_df3.to_csv(r"C:\Users\kmoudgalya\Desktop\AI PRD\Sample\check_json_output5.csv")

## Make a new ipynb, modify the approach.
## Instead of doing this check post the execution, we need a check during execution
## for example, count the number of occurences timestamp in a single cell, make sure that 
## all of those are covered and then move on to the next cell, 
## once done, do manual checks for 10 columns

## do this for all columns, as in run the loop, if it fails, mention the column and the count